src/vsr/replica.zig

```diff
@@ -2951,108 +2936,111 @@ pub fn ReplicaType(
             return message.ref();
         }
 
-        /// Discards uncommitted headers during a view change before the new leader starts the view.
+        /// Returns `replica.op + 1` when this replica (the new leader) participated in the last
+        /// `view_normal`. Otherwise returns the op of the lowest uncanonical message.
+        ///
+        /// An uncanonical message may have been removed/changed by a prior view.
+        ///
+        /// Consider these logs:
+        ///
+        ///   replica 0: 4, 5, 6b, 7b, 8b  (commit_min=6b, leader, status=normal, view=X)
+        ///   replica 1: 4, 5, 6b, --, --  (commit_min=5, follower, status=normal, view=X)
+        ///   replica 2: 4, 5, 6a, --, 8b  (view<X)
+        ///
+        /// 1. Replica 0 crashes immediately after committing 6b.
+        /// 2. Replicas 1 and 2 must determine the new chain HEAD.
+        /// 3. 8b is discarded due to the gap in 7.
+        /// 4. To distinguish between 6a and 6b (and safely discard 6a), the new leader trusts ops
+        ///    from the DVC(s) with the greatest `view_normal`.
+        fn op_uncanonical_min(self: *const Self, view_normal_canonical: u64) usize {
```

Comment:
:thinking: My working definition for uncanonical headers is:

> An uncanonical message _may have_ been removed/changed by a prior view.

i.e. the new leader may decide to keep some "uncanonical" headers. By that definition I think the naming is correct. Are you using "canonical" to mean "all headers part of the new view, according to the leader at view start"?

Reply:
> Are you using "canonical" to mean "all headers part of the new view, according to the leader at view start"?

Yes. And these we know for sure to be canonical.

> My working definition for uncanonical headers is:

>> An uncanonical message may have been removed/changed by a prior view.

Yes, but I'm thinking that it could also just as well be canonical on another replica. So it's more a question of uncertainty. Not knowing for sure whether it is canonical or not. But we can't say for sure that it is uncanonical.

Reply:
i.e. Like the opposite of a bloom filter, it will tell you for sure if something is not in a set, but it can't tell you if something is in a set. The opposite of that. 😁

Reply:
Gotcha, we were using different definitions for "canonical headers" then. I was thinking of it in terms of the new leader's viewpoint, with the uncertainty built in. Phrased in the positive, my definition could have been "A canonical header was part of the latest view_normal".

(Even the "canonical" headers (according to this definition) may still be discarded if there is a gap across all DVCs.)

Reply:
What if we define "the highest view_normal before now" to be the "base view" (or similar — "base" is probably too general a term). Then `op_canonical_max()` can be named (e.g.) `op_base_max()`.

Reply:
> Phrased in the positive, my definition could have been "A canonical header was part of the latest view_normal".

I was actually thinking this too, and also that these are the same, since the headers from the latest `view_normal` will be installed by the new primary and be part of the `start_view`.

> Even the "canonical" headers (according to this definition) may still be discarded if there is a gap across all DVCs.

Yes, agreed.

Reply:
> I was actually thinking this too, and also that these are the same, since the headers from the latest view_normal will be installed by the new primary and be part of the start_view.

They will be installed, but will not necessarily be part of the `start_view`. A subset of the canonical headers (meaning here: headers from the latest `view_normal`) may be removed before the `start_view` (by `discard_uncommitted_headers`) if there is a gap in them (that the DVC quorum could not fill/repair). So I don't think they _are_ the same thing.

(I think this might be the root of our confusion here, so lmk if I should elaborate).

Reply:
Ah, I was not being precise enough. I was thinking they are the same _ceteris parabus_, but of course if we discard because of gaps then they will differ.

But canonical is all w.r.t. to other possible candidate headers for the slots coming in from older views.

Reply:
That's how I was thinking of _canonical_. All about choosing which headers trump others, not so much about what will end up in the `start_view`.

Reply:
Gotcha; that makes sense! I will expand on this in the code — it is a useful term to have when explaining/describing view change precedence.

Then for renaming `op_canonical_max()`, something like `leader_op_canonical_max()` (since it is only called by the new leader)?

Reply:
i.e. canonical headers may be uncommitted or committed, so they may still be discarded. They're canonical in the sense that they're ratified by being part of the DVC messages from the latest `view_normal`, and they help to decide which other headers are allowed in via hash chain connection.

Happy to think of another name also. I don't know if saying "base headers" is as clear, although "base" does mean a starting point. We need to capture the sense of "authoritative" or "approved" or "quorum". Maybe "quorum headers"?

Reply:
Let's keep it short with `op_canonical_max()` — we also have other methods only used by the primary, but it's fine if we don't prefix them.

It would be great to rename all the `leader` references to `primary` one day.